src/graph/impl/KokkosGraph_Distance2Color_impl.hpp

```diff
@@ -737,6 +748,31 @@ class GraphColorDistance2
           lno_t vertsPerThread = 1;
           lno_t workBatches = (currentWork + vertsPerThread - 1) / vertsPerThread;
           timer.reset();
+          //if still using this color set, refresh forbidden.
+          //This avoids using too many colors, by relying on forbidden from before previous conflict resolution (which is now stale).
+          //Refreshing forbidden before conflict resolution ensures that previously-colored vertices do not get recolored.
+          switch(batch)
+          {
+            case 1:
+              Kokkos::parallel_for("NB D2 Forbidden", range_policy_type(0, numCols),
+                  NB_RefreshForbidden<1>(colorBase, forbidden, colors_out, this->t_xadj, this->t_adj, numVerts));
+              break;
+            case 2:
+              Kokkos::parallel_for("NB D2 Forbidden", range_policy_type(0, numCols),
+                  NB_RefreshForbidden<2>(colorBase, forbidden, colors_out, this->t_xadj, this->t_adj, numVerts));
+              break;
+            case 4:
+              Kokkos::parallel_for("NB D2 Forbidden", range_policy_type(0, numCols),
+                  NB_RefreshForbidden<4>(colorBase, forbidden, colors_out, this->t_xadj, this->t_adj, numVerts));
+              break;
+            case 8:
+              Kokkos::parallel_for("NB D2 Forbidden", range_policy_type(0, numCols),
+                  NB_RefreshForbidden<8>(colorBase, forbidden, colors_out, this->t_xadj, this->t_adj, numVerts));
+              break;
+            default:;
+          }
+          forbiddenTime += timer.seconds();
+          timer.reset();
```

Comment:
@brian-kelley: Is a fence needed between refreshing forbidden and coloring the nodes?

Reply:
No, all the functors execute on the same execution space instance so no fence is needed.